ocaml/runtime/minor_gc.c

```diff
@@ -130,10 +130,20 @@ static void reset_table (struct generic_table *tbl)
   tbl->base = tbl->ptr = tbl->threshold = tbl->limit = tbl->end = NULL;
 }
 
-static void clear_table (struct generic_table *tbl)
+static void clear_table (struct generic_table *tbl,
+                         asize_t element_size,
+                         const char* name)
 {
+  asize_t maxsz = Caml_state->minor_heap_wsz;
+  if (tbl->size <= maxsz) {
     tbl->ptr = tbl->base;
     tbl->limit = tbl->threshold;
+  } else {
+    caml_gc_message (0x08, "Shrinking %s to %ldk bytes\n",
+                     name,
+                     (long)((maxsz * element_size) / 1024));
+    alloc_generic_table(tbl, Caml_state->minor_heap_wsz, 256, element_size);
```

Comment:
Was this supposed to match the initial size of these tables?  If so, it looks like a division by eight is needed in the second argument (see realloc_generic_table).

Reply:
No, it's supposed to be larger. The ref_table being allowed to grow is a useful feature (especially in the absence of safepoints) - a larger-than-default ref table means fewer forced minor GCs. There just needs to be some bound on _how_ large.